osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs

```diff
@@ -107,9 +113,78 @@ public void AddRotation(float angle)
             }
 
             currentRotation += angle;
+
             // rate has to be applied each frame, because it's not guaranteed to be constant throughout playback
             // (see: ModTimeRamp)
-            drawableSpinner.Result.RateAdjustedRotation += (float)(Math.Abs(angle) * (gameplayClock?.GetTrueGameplayRate() ?? Clock.Rate));
+            double time = gameplayClock?.CurrentTime ?? Clock.CurrentTime;
+            double rate = gameplayClock?.GetTrueGameplayRate() ?? Clock.Rate;
+            float rateAdjustedRotation = (float)(angle * rate);
+
+            if (rate > 0)
+            {
+                int direction = Math.Sign(rateAdjustedRotation);
+
+                if (direction != 0 && direction != currentDirection)
+                {
+                    turningPoints.Add((time, currentDelta));
+                    currentDelta = 0;
+                }
+
+                currentDirection = direction;
+            }
+            else if (turningPoints.Count > 0 && time <= turningPoints[^1].time)
```

Comment:
Also assuming this is doing rewind, should it not be looping while `time <= turningPoints[^1].time`?

Reply:
Fair point

Reply:
I'd rewrite this section like this, for better readability:

```diff
diff --git a/osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs b/osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs
index f580635565..3f6ba3f72e 100644
--- a/osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs
+++ b/osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs
@@ -119,7 +119,18 @@ public void AddRotation(float angle)
             double rate = gameplayClock?.GetTrueGameplayRate() ?? Clock.Rate;
             float rateAdjustedRotation = (float)(angle * rate);
 
-            if (rate > 0)
+            bool isRewinding = rate > 0;
+
+            if (isRewinding)
+            {
+                while (turningPoints.Count > 0 && time <= turningPoints[^1].time)
+                {
+                    currentDelta = turningPoints[^1].delta;
+                    currentDirection = Math.Sign(currentDelta);
+                    turningPoints.RemoveAt(turningPoints.Count - 1);
+                }
+            }
+            else
             {
                 int direction = Math.Sign(rateAdjustedRotation);
 
@@ -131,12 +142,6 @@ public void AddRotation(float angle)
 
                 currentDirection = direction;
             }
-            else if (turningPoints.Count > 0 && time <= turningPoints[^1].time)
-            {
-                currentDelta = turningPoints[^1].delta;
-                currentDirection = Math.Sign(currentDelta);
-                turningPoints.RemoveAt(turningPoints.Count - 1);
-            }
 
             currentDelta += rateAdjustedRotation;
             drawableSpinner.Result.RateAdjustedRotation = computeEffectiveRotation();

```